### PR TITLE
JWT Cookie Authentication

### PR DESCRIPTION
**Pull request by bastihav**
# Description

Implements #186, #289

I would like to wait for the test branch before merging and testing this.
Should not influence any production code though.

To run this, webpack will need to spin up a reverse proxy with TLS encryption. Use the following command to generate a localhost certificate (browser will tell you that it is untrusted)

`openssl req -x509 -nodes -days 1024 -newkey rsa:2048 -keyout certs/localhost.key -out certs/localhost.crt -extensions 'v3_req' -subj '/C=DE/ST=NRW/L=Paderborn/CN=localhost'`

Caution: highly WIP. Currently using different certs to debug pipeline:
```
      - name: Create CA certificate
        run: openssl req -x509 -newkey rsa:4096 -days 365 -nodes -keyout certs/ca-key.pem -out certs/ca-cert.pem -subj '/C=DE/ST=NRW/L=Paderborn/CN=ca-localhost'
      - name: Create certificate request
        run: openssl req -newkey rsa:4096 -nodes -keyout certs/server-key.pem -out certs/server-req.pem -subj '/C=DE/ST=NRW/L=Paderborn/CN=localhost'
      - name: Sign certificate request
        run: openssl x509 -req -in certs/server-req.pem -days 60 -CA certs/ca-cert.pem -CAkey certs/ca-key.pem -CAcreateserial -out certs/server-cert.pem
```

## Reason for this PR
- See #289 for more details

## Changes in this PR
- remove login data from store, change authentication to use httpOnly cookies, add new login and logout methods

## Type of change (remove all that don't apply)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Manually tested all use cases this could alter

**Test Configuration**:
- OS: Windows
- Browser: Chrome

- Frontend: (remove all that don't apply)
  - [X] Development build

- Backend: (remove all that don't apply)
  - [X] deployed experimental build

# Checklist: (remove all that don't apply)

- [x] I added corresponding E2E tests (especially for bugfixes)
- [x] I added corresponding unit tests (if API code)
- [X] I have performed a self-review of my own code
- [X] My changes generate no new linting warnings or console warnings
- [x] I have updated the package.json version if this is a new release candidate (remove is it is not)
- [x] I have updated the CHANGELOG.md to include any changes made in this PR (add WIP to the top, if there is none already)
- [x] I have changed the API endpoint back to development

### CODE REVIEW

File: src/api/AuthenticationManagement.ts [hunk 1]
```diff
@@ -33,29 +39,173 @@ export default class AuthenticationManagement extends Common {
...
+
+            store.commit(MutationTypes.SET_LOGGEDIN, true);
+
+            const userManagement = new UserManagement();
+            const handler = new GenericResponseHandler();
+            const response = await userManagement.getSpecificUser(username);
+            const user = handler.handleResponse(response);
+            store.commit(MutationTypes.SET_USER, user);
+            // // set role after user, because the navbar is loaded as soon as the role is set.
+            // store.commit(MutationTypes.SET_ROLE, intermediateResult.returnValue);
```
Comment (tobias-wiese) [inline src/api/AuthenticationManagement.ts:130]: Can this be deleted, as you set a user object?

File: src/api/AuthenticationManagement.ts [hunk 2]
```diff
@@ -33,29 +39,173 @@ export default class AuthenticationManagement extends Common {
...
+        if (result.returnValue) {
+            const store = useStore();
+
+            store.commit(MutationTypes.SET_LOGGEDIN, true);
+
+            const userManagement = new UserManagement();
+            const handler = new GenericResponseHandler();
+            const response = await userManagement.getSpecificUser(username);
+            const user = handler.handleResponse(response);
+            store.commit(MutationTypes.SET_USER, user);
```
Comment (tobias-wiese) [inline src/api/AuthenticationManagement.ts:128]: What if the API call fails? Then we should not commit the user, as we do not have one, should we?

File: src/api/AuthenticationManagement.ts [hunk 3]
```diff
@@ -33,29 +39,173 @@ export default class AuthenticationManagement extends Common {
...
+
+        if (result.returnValue) {
+            const store = useStore();
+            store.commit(MutationTypes.SET_LOGGEDIN, true);
+
+            const userManagement = new UserManagement();
+            const handler = new GenericResponseHandler();
+            const response = await userManagement.getSpecificUser(loginData.username);
+            const user = handler.handleResponse(response);
+            store.commit(MutationTypes.SET_USER, user);
```
Comment (tobias-wiese) [inline src/api/AuthenticationManagement.ts:206]: see above

File: src/components/account/list/UserRow.vue [hunk 4]
```diff
@@ -38,7 +38,7 @@
                 <div class="flex-col items-baseline" :class="[isStudent ? 'flex' : 'hidden']">
                     <div class="leading-5 text-blue-900 ml-1 mb-1">{{ student.matriculationId }}</div>
                     <div class="hidden sm:flex items-center leading-5 text-gray-500">
-                        <span class="mr-2 fa-stack text-xs" style="font-size: 0.63em">
+                        <span class="mr-2 fa-stack text-xs" style="font-size: 0.63em;">
```
Comment (tobias-wiese) [inline src/components/account/list/UserRow.vue:41]:
```suggestion
                        <span class="mr-2 fa-stack text-xs" style="font-size: 0.63em">
```

is this semicolon intended? Just dismiss this suggestion if it is

File: vue.config.js [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
 
 // do not add trailing forward slash
-process.env.VUE_APP_API_BASE_URL =
-    process.env.NODE_ENV === "production"
-        ? "https://uc4.cs.uni-paderborn.de/api/production"
-        : "https://uc4.cs.uni-paderborn.de/api/develop";
+process.env.VUE_APP_API_BASE_URL = process.env.NODE_ENV === "production" ? "https://uc4.cs.uni-paderborn.de/api/production" : "/api";
+
+const endpoint = "https://uc4.cs.uni-paderborn.de/api/experimental/";
+// const endpoint = "https://uc4.cs.uni-paderborn.de/api/development/"
```
Comment (tobias-wiese) [inline vue.config.js:8]: For the merge on develop, we should use the endpoint "develop". (As soon as the changes are present there)

Comment (bastihav) [reply vue.config.js:8]: yes, see checklist in PR template

**Review by tobias-wiese — CHANGES_REQUESTED** [review]
We get the following vue warning for the unsaved changes and delete modals on the two existent forms
```Extraneous non-props attributes (id) were passed to component but could not be automatically inherited because component renders fragment or text root nodes. ``` 

(optional) We may want to hide the ```Das Cookie "login" wurde abgelehnt, da es bereits abgelaufen ist.``` warning that appears when logging  out.

**Comment by bastihav** [comment]
> (optional) We may want to hide the `Das Cookie "login" wurde abgelehnt, da es bereits abgelaufen ist.` warning that appears when logging out.

Not possible

**Comment by tobias-wiese** [comment]
- Relogin seems not to be implemented on about page (Login -> navigate to "/about" -> refresh -> not logged in any more).
- Application crashes on refreshing the course or account form in editmode (Chrome: ```Uncaught SyntaxError: Unexpected token '<'``` and Firefox: ```Das Skript von "https://localhost:8080/editAccount/js/chunk-vendors.js" wurde geladen, obwohl sein MIME-Typ ("text/html") kein für JavaScript gültiger MIME-Typ ist.```).

**Comment by bastihav** [comment]
I do not know how to update our unit tests to work with JWT cookies. Afaik jest does not really support http-only cookies. I did not succeed in mocking the cookies and still sending them with every request.

We could completely mock the backend for our unit tests, but this is a huge overhead.

E2E tests _should_ cover all cases of our unit testing anyways, so I would suggest removing the unit tests for now and maybe revisiting the topic at another time. 

If anyone knows how to get jest to behave and save + send http-only secure cookies, feel free to implement that.

Please let me know what you think

**Comment by tobias-wiese** [comment]
> 
> 
> I do not know how to update our unit tests to work with JWT cookies. Afaik jest does not really support http-only cookies. I did not succeed in mocking the cookies and still sending them with every request.
> 
> We could completely mock the backend for our unit tests, but this is a huge overhead.
> 
> E2E tests _should_ cover all cases of our unit testing anyways, so I would suggest removing the unit tests for now and maybe revisiting the topic at another time.
> 
> If anyone knows how to get jest to behave and save + send http-only secure cookies, feel free to implement that.
> 
> Please let me know what you think

I don't think that the overhead is worth it. As you mention, the E2E should cover all positive cases. However, we then need to adapt our workflow as we could not test the API functionality _before_ using them for the UI implementation.

**Comment by bastihav** [comment]
> * Application crashes on refreshing the course or account form in editmode (Chrome: `Uncaught SyntaxError: Unexpected token '<'` and Firefox: `Das Skript von "https://localhost:8080/editAccount/js/chunk-vendors.js" wurde geladen, obwohl sein MIME-Typ ("text/html") kein für JavaScript gültiger MIME-Typ ist.`).

This is just a problem with development builds. I am not sure how to fix this with webpack, but it works in production.

File: src/views/lecturer/EditCreateCourseForm.vue [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+                    alert("Course not found");
+                } else {
+                    course.value = new CourseEntity(result);
+                    initialCourseState = JSON.parse(JSON.stringify(course.value));
+                }
+                busy.value = false;
+            }
+
+            let hasInput = computed(() => {
+                // TODO not tested yet (too lazy to start intellij)
```
Comment (theovier) [inline src/views/lecturer/EditCreateCourseForm.vue:208]: 🧐

**Review by theovier — APPROVED** [review]

**Review by tobias-wiese — APPROVED** [review]

**Review by janniclas — COMMENTED** [review]
I just added a few general remarks and thoughts to the code, which came up when I read it. These are meant for u as feedback of what I thought when reading it, so feel free to do with these comments as u see fit. 
In general I think the code quality is very high and I didn't see any obvious bugs. 

Few further remarks: 

- on the About page the html <a> tag should  probably use the rel attribute (https://web.dev/external-anchors-use-rel-noopener/)

- if u are loged in, open another tab, visit the same page, logout on one of these pages, the other isn't redirected to the login page. The information isn't loaded anymore so that's fine but I would expect a redirect. Also on the Settings page the change password component is still displayed and can actualy be used to relogin by providing a password which isn't expected behaviour I believe.

- lot's of unused imports, maybe clean that up at some point :)

**Review by janniclas — APPROVED** [review]

**Pull request merged**